File: 0\1_knapsack.dart
import 'dart:math';

void main() 
{
  List<int> weights = [2,3,4,5];
  List<int> values = [3,4,5,6];
  int capacity = 8;

  print("Maximum value: ${knapsack(weights, values, capacity)}");
}

int knapsack(List<int> weights, List<int> values, int capacity) 
{
  int n = weights.length;
  List<int> dp = List.filled(capacity + 1, 0); 

  for (int a = 0; a < n; a++) 
  {
    for (int w = capacity; w >= weights[a]; w--) 
    {
      dp[w] = max(dp[w], dp[w - weights[a]] + values[a]);
    }
  }

  return dp[capacity];
}

File: DP/0\1_knapsack.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int knapsack(vector<int>& weights, vector<int>& values, int capacity) 
{
    int n = weights.size();
    vector<int> dp(capacity + 1, 0);

    for (int a = 0; a < n; a++) 
    {
        for (int w = capacity; w >= weights[a]; w--) 
        {
            dp[w] = max(dp[w], dp[w - weights[a]] + values[a]);
        }
    }

    return dp[capacity];
}

int main() 
{
    vector<int> weights = {2, 3, 4, 5};
    vector<int> values = {3, 4, 5, 6};
    int capacity = 8;

    cout << "Maximum value: " << knapsack(weights, values, capacity) << endl;

    return 0;
}
